OSOE-49: ⭐ Replace Windows-only code in UI Testing Toolbox and Fix Lombiq.Analyzers on Linux

sarahelsaig opened the pull request:
OSOE-49

sarahelsaig committed:
Return submodules to dev.

sarahelsaig committed:
submodules

sarahelsaig committed:
Merge remote-tracking branch 'origin/dev' into issue/OSOE-49

sarahelsaig committed:
submodules

sarahelsaig committed:
Merge remote-tracking branch 'origin/dev' into issue/OSOE-49

sarahelsaig committed:
Fix gitattributes.

sarahelsaig committed:
Use NVM for NPM if possible, and bug fix.

sarahelsaig committed:
submodules

sarahelsaig committed:
WebDriverFactory

sarahelsaig committed:
Code styling.

sarahelsaig committed:
More code styling.

sarahelsaig committed:
Fix DateTime culture problem.

The same way you see in ClosedXML tests.

sarahelsaig committed:
Line endings stuff.

sarahelsaig committed:
Added Linux instructions.

sarahelsaig committed:
Don't suggest going for node v14 by default.

sarahelsaig committed:
Replace deprecated and broken package.

sarahelsaig committed:
Merge remote-tracking branch 'origin/dev' into issue/OSOE-49

sarahelsaig committed:
Merge remote-tracking branch 'origin/dev' into issue/OSOE-49

sarahelsaig committed:
submodules

sarahelsaig committed:
submodule

sarahelsaig committed:
Fix webroot argument that somehow corrupted the following environment argument.

sarahelsaig committed:
Merge remote-tracking branch 'origin/dev' into issue/OSOE-49

sarahelsaig committed:
Disable UseSqlServer in BasicTests.

There is no point because there are separate SQL Server specific tests.

sarahelsaig committed:
submodule

sarahelsaig committed:
Shared TestConfiguration.json and more docs.

sarahelsaig committed:
Fix host path problem with tilde.

sarahelsaig committed:
Fix file access problem during SQL backup.

sarahelsaig committed:
Code styling.

sarahelsaig committed:
submodule

sarahelsaig committed:
Merge remote-tracking branch 'origin/dev' into issue/OSOE-49

sarahelsaig committed:
Update submodule.

sarahelsaig committed:
Create testing docker image + readme (I've committed it to the wrong branch).

(cherry picked from commit 4800458012235db97f47e69020efaa86adef357c)

sarahelsaig committed:
Call clean first.

sarahelsaig committed:
Improve docker practices.

sarahelsaig committed:
Add missing NPM package reference

that prevented first time successful build.

sarahelsaig committed:
Add missing runtime and expose ports from the inside too.

sarahelsaig committed:
dev-cert and other cleanup

sarahelsaig committed:
Improve readme.

sarahelsaig committed:
Merge branch 'issue/OSOE-60' into issue/OSOE-49

# Conflicts:
#	src/Modules/Lombiq.ChartJs
#	src/Modules/Lombiq.DataTables
#	src/Modules/Lombiq.Hosting.Tenants
#	src/Modules/Lombiq.Privacy
#	src/Modules/Lombiq.UIKit
#	test/Lombiq.UITestingToolbox
#	tools/Lombiq.Analyzers

sarahelsaig committed:
Update Lombiq.UITestingToolbox.

sarahelsaig committed:
Update readme.

sarahelsaig committed:
Update/merge UI Testing Toolbox.

sarahelsaig committed:
solution fixes

sarahelsaig committed:
dos2unix for windows friends

sarahelsaig committed:
Rewrite to use "mcr.microsoft.com/dotnet/aspnet:6.0" instead.

sarahelsaig committed:
dockerfile bug fix

sarahelsaig committed:
Update submodules.

sarahelsaig committed:
Merge remote-tracking branch 'origin/dev' into issue/OSOE-49

# Conflicts:
#	Lombiq.OSOCE.sln

sarahelsaig committed:
Fix namespace.

sarahelsaig committed:
submodule

sarahelsaig committed:
Merge remote-tracking branch 'origin/dev' into issue/OSOE-49

sarahelsaig committed:
Renormalize.

sarahelsaig committed:
Fix Linux part of Configuration.md.

sarahelsaig committed:
Update Lombiq.Analyzers.

sarahelsaig committed:
Fix ruleset file case sensitivity.

sarahelsaig committed:
Renamed rulesets to all lower case.

sarahelsaig committed:
Merge remote-tracking branch 'origin/dev' into issue/OSOE-49

# Conflicts:
#	Directory.Build.props
#	src/Modules/Lombiq.Privacy
#	test/Lombiq.UITestingToolbox

sarahelsaig committed:
Add github actions.

sarahelsaig committed:
fix trigger branch name

sarahelsaig committed:
Clean up action.

sarahelsaig committed:
Rename parameter name that hides local variable.

sarahelsaig committed:
Enforce Windows newlines for C# files to avoid false positives with IDE55 warning.

sarahelsaig committed:
fix error CA1822: Member 'PaginateAsync' does not access instance data and can be marked as static

sarahelsaig committed:
inconsequential change

sarahelsaig committed:
Install pnpm in github action.

sarahelsaig committed:
Add action yml file to the solution items.

sarahelsaig committed:
Fix missing inner exception for fast fail setup.

sarahelsaig committed:
Improve CreateDriverAsync

sarahelsaig committed:
dummy commit

sarahelsaig committed:
dummy commit

sarahelsaig committed:
Make the Chrome driver version picking more forgiving.

sarahelsaig committed:
Code styling.

sarahelsaig committed:
Use version api uri.

sarahelsaig committed:
update submodules

sarahelsaig committed:
Merge remote-tracking branch 'origin/dev' into issue/OSOE-49

# Conflicts:
#	Lombiq.OSOCE.sln
#	src/Utilities/Lombiq.Gulp.Extensions

sarahelsaig committed:
renormalize

sarahelsaig committed:
Add Lombiq.HelpfulLibraries.Cli

sarahelsaig committed:
Refactor GetVersion to actually work correctly on Windows.

sarahelsaig committed:
Update submodules from origin/dev

sarahelsaig committed:
Merge remote-tracking branch 'origin/dev' into issue/OSOE-49

sarahelsaig committed:
Update gitattributes in Lombiq.BaseTheme.

sarahelsaig committed:
Fix comment in gitattributes.

sarahelsaig committed:
Do dotnet clean first.

sarahelsaig committed:
nope

sarahelsaig committed:
Whitelist IPv6 localhost.

sarahelsaig committed:
Resolve localhosts.

sarahelsaig committed:
Add disable-dev-shm-usage switch.

sarahelsaig committed:
Check out this CHROMEWEBDRIVER variable.

sarahelsaig committed:
Don't download chrome driver if it's already installed.

sarahelsaig committed:
hosts trickery

sarahelsaig committed:
Simplify

sarahelsaig committed:
Force headless mode if we are in Linux without a working graphical environment.

sarahelsaig committed:
Test every project in solution.

sarahelsaig committed:
Don't try to run Lombiq.Tests.UI.csproj

sarahelsaig committed:
Filter out trash logs.

Using workaround based on suggestion at https://stackoverflow.com/questions/64914726/selenium-powershell-linux-connection-refused-ffff127-0-0-139343-localhos#comment127401794_64914726

sarahelsaig committed:
submodule

sarahelsaig committed:
Try not to obfuscate the test result.

sarahelsaig committed:
Fix truncated smtp4dev error reporting.

sarahelsaig committed:
Unusing.

sarahelsaig committed:
Add SQL Server Docker

sarahelsaig committed:
Fix syntax error

sarahelsaig committed:
Group outputs and other cleanup.

sarahelsaig committed:
Set up Azurite via Docker

sarahelsaig committed:
Fix Azurite installation so it doesn't stall.

sarahelsaig committed:
Use DotnetCli.

sarahelsaig committed:
Organize yml

sarahelsaig committed:
Probe LOMBIQ_SHARED_TEST_CONFIGURATION.

sarahelsaig committed:
Code styling.

sarahelsaig committed:
Replace direct chown call with more reliable bash call.

sarahelsaig committed:
Add missing reference.

sarahelsaig committed:
Rewrite docker SQL snapshot behavior.

sarahelsaig committed:
Code styling.

sarahelsaig committed:
Unusing

sarahelsaig committed:
Extract GitHub-specific TestConfig.json

sarahelsaig committed:
Reset ownership for the backup copied into docker.

sarahelsaig committed:
Code styling.

sarahelsaig committed:
Bug fix.

sarahelsaig committed:
Bug fix.

sarahelsaig committed:
Bug fix.

sarahelsaig committed:
Code cleanup after Atata upgrade.

sarahelsaig committed:
Add default ProgramCli configuration and document it.

sarahelsaig committed:
update analyzers

sarahelsaig committed:
Merge remote-tracking branch 'origin/dev' into issue/OSOE-49

sarahelsaig committed:
yml do 2nd try

sarahelsaig committed:
Code styling.

sarahelsaig committed:
Upgrade SDK and XUnit runner.

sarahelsaig committed:
Obsolete GetAndEmptyBrowserLogAsync, replace with GetAndEmptyBrowserLog

sarahelsaig committed:
Temporarily disable emailing test.

sarahelsaig committed:
Fix monkey testing after the log update.

sarahelsaig committed:
Fix test.

sarahelsaig committed:
Prevent lockup.

sarahelsaig committed:
Code styling.

sarahelsaig committed:
Upgrade rnwood.smtp4dev and re-enable test.

sarahelsaig committed:
smtp4dev repro

sarahelsaig committed:
Exclude only GitHub Actions from the email test.

sarahelsaig committed:
Fix excel export unit test.

sarahelsaig committed:
Add missing NPM reference.

sarahelsaig committed:
Remove docker file that didn't end up working.

sarahelsaig committed:
update gitattributes

Piedone commented:
Merge from dev.

sarahelsaig committed:
update submodule

sarahelsaig committed:
Merge remote-tracking branch 'origin/dev' into issue/OSOE-49

# Conflicts:
#	Lombiq.OSOCE.sln
#	src/Lombiq.OSOCE.Web/Lombiq.OSOCE.Web.csproj
#	src/Modules/Lombiq.ChartJs

sarahelsaig commented:
ok, merged

Piedone committed:
Merge remote-tracking branch 'origin/dev' into issue/OSOE-49

# Conflicts:
#	tools/Lombiq.Analyzers

Piedone commented:
And fix the build errors too :).

sarahelsaig committed:
Post merge fixup.

sarahelsaig committed:
Merge branch 'issue/OSOE-49' of https://github.com/Lombiq/Open-Source-Orchard-Core-Extensions into issue/OSOE-49

Piedone commented on .github/workflows/analyze_and_test.yml:
```diff
@@ -0,0 +1,50 @@
+name: Pull Request - Code Analysis and Unit Tests
+on:
+  pull_request:
+    branches: [ dev ]
```
We sometimes open longer-living feature branches with multiple PRs to them, so we don't need this restriction.

Piedone commented on .github/workflows/analyze_and_test.yml:
```diff
@@ -0,0 +1,50 @@
+name: Pull Request - Code Analysis and Unit Tests
```
```suggestion
name: Pull Request - Build and Test
```

Piedone commented on .github/workflows/analyze_and_test.yml:
```diff
@@ -0,0 +1,50 @@
+jobs:
+  build_test:
+    runs-on: ubuntu-latest
+    name: Build & Test
```
Just to be in line with TC:

```suggestion
    name: Build and Test
```

sarahelsaig committed:
This should make the GitHub CI less flaky.

Piedone commented on .github/workflows/analyze_and_test.yml:
```diff
@@ -0,0 +1,51 @@
+    - name: Static Code Analysis
+      run: |
+        npm install pnpm -g &&
+        dotnet build src/Utilities/Lombiq.Npm.Targets/Lombiq.Npm.Targets.csproj
```
Wouldn't project references to NPM Targets, just as to Gulp Extensions, fix this too?

sarahelsaig replied:
Nope, because even this didn't fix it. Had to add a retry logic to make it reliable in the end. The affected projects already reference Lombiq.Npm.Targets so I have no idea why it's like this.

sarahelsaig replied:
Just needed to pre-build Gulp Extensions. Weird, becuase based on the project dependency trees that should already happen by that point.

Piedone commented on .github/workflows/analyze_and_test.yml:
```diff
@@ -0,0 +1,51 @@
+      run: |
+        npm install pnpm -g &&
+        dotnet build src/Utilities/Lombiq.Npm.Targets/Lombiq.Npm.Targets.csproj
+        dotnet build Lombiq.OSOCE.sln --no-incremental -warnaserror /p:TreatWarningsAsErrors=true /p:RunAnalyzersDuringBuild=true -nologo -consoleLoggerParameters:NoSummary -verbosity:quiet
```
`--no-incremental` shouldn't be needed on a single-use build agent.

Also, better to build like we'd run it in prod: `--configuration Release`. I don't really understand why there's a second Release build after this.

And why quiet?

sarahelsaig replied:
For some reason I thought it has to be in Debug. If it can be Release, then indeed there is no need for a second build.

It's quiet to eliminate irrelevant outputs. Really, all we care about is if there is some build error and if there isn't then there is no point outputting [this garbage information](https://github.com/Lombiq/Open-Source-Orchard-Core-Extensions/runs/6430683041?check_suite_focus=true#step:6:12).

0liver replied:
Good point about the `quiet` verbosity level. We might want to adapt that as the default in [our GitHub publish action](https://github.com/Lombiq/GitHub-Actions/blob/dev/.github/workflows/publish.yml#L12), as well.

Piedone replied:
For Shouldly-using unit tests a Debug publish would be somewhat nicer but this doesn't matter much.

Piedone commented on .github/workflows/analyze_and_test.yml:
```diff
@@ -0,0 +1,51 @@
+    - name: Build
+      run: |
+        dotnet build --configuration Release
+    - name: Set up SQL Server via Docker
```
Wow, nice!

Piedone commented on .github/TestConfig.json:
```diff
@@ -0,0 +1,10 @@
+{
+  "Lombiq_Tests_UI": {
```
It would be a bit nicer to keep these configs in environment variables, see `TestConfigurationManager` (you can make more code configs configurable from the env, just add more `TestConfigurationManager` calls with defaults keeping the current defaults).

sarahelsaig replied:
Could you explain why you think that would be nicer? I think adding some complicated variable like `$Lombiq_Tests_UI__SqlServerDatabaseConfiguration__ConnectionStringTemplate` would be less readable and it would be even worse if we had more settings, though just 2 is borderline.

Piedone replied:
I learned about the central config file feature in the meantime. So to confirm, this file is used from all test projects and isn't copied around, right?

sarahelsaig replied:
Yes.

Piedone replied:
Good then.

Piedone commented on .github/workflows/analyze_and_test.yml:
```diff
@@ -0,0 +1,51 @@
+      run: |
+        docker pull mcr.microsoft.com/azure-storage/azurite &&
+        docker run --name azurite -d -p 10000:10000 mcr.microsoft.com/azure-storage/azurite azurite-blob --blobHost 0.0.0.0 --blobPort 10000
+    - name: Unit Tests
```
Since these are also UI tests:

```suggestion
    - name: Tests
```

Piedone commented on .github/workflows/analyze_and_test.yml:
```diff
@@ -0,0 +1,51 @@
+        grep -v 'Lombiq.Tests.csproj' |
+        sed -e 's/\\/\//g' -e '$atrue' -e 's/\(.*\)/echo "::group::\1" \&\&\ndotnet test -c Release --no-restore --no-build '"'\\1'"' \&\&\necho "::endgroup::" \&\&/' |
+        bash && (echo SUCCESS > .ui_test_result) || (echo FAIL > .ui_test_result)) | grep -v '^Connection refused \[::ffff:127.0.0.1\]';
+        grep SUCCESS .ui_test_result
```
Make failure dumps available as an artifact: https://github.com/OrchardCMS/OrchardCore/pull/11194/files#diff-1153a5676594ad20522a7686e7e94a258891fe87611bafcd10ec0f3943c219b8

sarahelsaig replied:
life is kill 😭 
![image](https://user-images.githubusercontent.com/4007293/168423831-1933a20d-ee45-408a-8e4e-6da3d29c1d27.png)

0liver replied:
At least the error description is very verbose 😁

sarahelsaig replied:
The last line makes me think it's so verbose because people were opening change request issues against this behavior.

Piedone replied:
\> when were you when Dávid got upset?
\> i was sat at home, reviewing code, chilling to synthwave when e-mail with Dávid's comment
\> 'contains the following character: colon :'
\> 'no'

sarahelsaig replied:
> chilling to synthwave

👍 👍 👍 👍

Piedone commented on .github/workflows/analyze_and_test.yml:
```diff
@@ -0,0 +1,51 @@
+        grep -v 'Lombiq.Tests.UI.csproj' |
+        grep -v 'Lombiq.Tests.csproj' |
+        sed -e 's/\\/\//g' -e '$atrue' -e 's/\(.*\)/echo "::group::\1" \&\&\ndotnet test -c Release --no-restore --no-build '"'\\1'"' \&\&\necho "::endgroup::" \&\&/' |
+        bash && (echo SUCCESS > .ui_test_result) || (echo FAIL > .ui_test_result)) | grep -v '^Connection refused \[::ffff:127.0.0.1\]';
```
What on earth is happening here? :D Also, you can filter tests from within `dotnet test` too: https://docs.microsoft.com/en-us/dotnet/core/testing/selective-unit-tests?pivots=mstest

sarahelsaig replied:
That filter is not helpful because of the GitHub specific tricks that go on here. The goal is to place every test into a separate collapsible section within the step. See the generated script when you run the code up to this line in bash:
```sh
$ dotnet sln list |
>         grep '.Tests.' |
>         grep -v 'Lombiq.Tests.UI.csproj' |
>         grep -v 'Lombiq.Tests.csproj' |
>         sed -e 's/\\/\//g' -e '$atrue' -e 's/\(.*\)/echo "::group::\1" \&\&\ndotnet test -c Release --no-restore --no-build '"'\\1'"' \&\&\necho "::endgroup::" \&\&/'

echo "::group::src/Modules/Lombiq.TrainingDemo/Tests/Lombiq.TrainingDemo.Tests/Lombiq.TrainingDemo.Tests.csproj" &&
dotnet test -c Release --no-restore --no-build 'src/Modules/Lombiq.TrainingDemo/Tests/Lombiq.TrainingDemo.Tests/Lombiq.TrainingDemo.Tests.csproj' &&
echo "::endgroup::" &&
echo "::group::test/Lombiq.UITestingToolbox/Lombiq.Tests.UI.Shortcuts/Lombiq.Tests.UI.Shortcuts.csproj" &&
dotnet test -c Release --no-restore --no-build 'test/Lombiq.UITestingToolbox/Lombiq.Tests.UI.Shortcuts/Lombiq.Tests.UI.Shortcuts.csproj' &&
echo "::endgroup::" &&
echo "::group::test/Lombiq.UITestingToolbox/Lombiq.Tests.UI.AppExtensions/Lombiq.Tests.UI.AppExtensions.csproj" &&
dotnet test -c Release --no-restore --no-build 'test/Lombiq.UITestingToolbox/Lombiq.Tests.UI.AppExtensions/Lombiq.Tests.UI.AppExtensions.csproj' &&
echo "::endgroup::" &&
echo "::group::test/Lombiq.UITestingToolbox/Lombiq.Tests.UI.Samples/Lombiq.Tests.UI.Samples.csproj" &&
dotnet test -c Release --no-restore --no-build 'test/Lombiq.UITestingToolbox/Lombiq.Tests.UI.Samples/Lombiq.Tests.UI.Samples.csproj' &&
echo "::endgroup::" &&
echo "::group::src/Libraries/Lombiq.HelpfulLibraries/Lombiq.HelpfulLibraries.Tests/Lombiq.HelpfulLibraries.Tests.csproj" &&
dotnet test -c Release --no-restore --no-build 'src/Libraries/Lombiq.HelpfulLibraries/Lombiq.HelpfulLibraries.Tests/Lombiq.HelpfulLibraries.Tests.csproj' &&
echo "::endgroup::" &&
echo "::group::src/Modules/Lombiq.DataTables/Lombiq.DataTables/Tests/Lombiq.DataTables.Tests/Lombiq.DataTables.Tests.csproj" &&
dotnet test -c Release --no-restore --no-build 'src/Modules/Lombiq.DataTables/Lombiq.DataTables/Tests/Lombiq.DataTables.Tests/Lombiq.DataTables.Tests.csproj' &&
echo "::endgroup::" &&
echo "::group::test/Lombiq.OSOCE.Tests.UI/Lombiq.OSOCE.Tests.UI.csproj" &&
dotnet test -c Release --no-restore --no-build 'test/Lombiq.OSOCE.Tests.UI/Lombiq.OSOCE.Tests.UI.csproj' &&
echo "::endgroup::" &&
echo "::group::src/Modules/Lombiq.DataTables/Lombiq.DataTables/Tests/Lombiq.DataTables.Tests.UI/Lombiq.DataTables.Tests.UI.csproj" &&
dotnet test -c Release --no-restore --no-build 'src/Modules/Lombiq.DataTables/Lombiq.DataTables/Tests/Lombiq.DataTables.Tests.UI/Lombiq.DataTables.Tests.UI.csproj' &&
echo "::endgroup::" &&
echo "::group::src/Modules/Lombiq.VueJs/Lombiq.VueJs.Tests.UI/Lombiq.VueJs.Tests.UI.csproj" &&
dotnet test -c Release --no-restore --no-build 'src/Modules/Lombiq.VueJs/Lombiq.VueJs.Tests.UI/Lombiq.VueJs.Tests.UI.csproj' &&
echo "::endgroup::" &&
echo "::group::src/Themes/Lombiq.BaseTheme/Lombiq.BaseTheme.Tests.UI/Lombiq.BaseTheme.Tests.UI.csproj" &&
dotnet test -c Release --no-restore --no-build 'src/Themes/Lombiq.BaseTheme/Lombiq.BaseTheme.Tests.UI/Lombiq.BaseTheme.Tests.UI.csproj' &&
echo "::endgroup::" &&
echo "::group::src/Modules/Lombiq.ChartJs/Lombiq.ChartJs.Tests.UI/Lombiq.ChartJs.Tests.UI.csproj" &&
dotnet test -c Release --no-restore --no-build 'src/Modules/Lombiq.ChartJs/Lombiq.ChartJs.Tests.UI/Lombiq.ChartJs.Tests.UI.csproj' &&
echo "::endgroup::" &&
true
```

The `echo "::group::name` and echo "::endgroup::" are the magic strings that let you achieve this clean effect:
![image](https://user-images.githubusercontent.com/4007293/168422248-09b35afd-c33b-4de1-8482-e3a7c0adce8e.png)

Also everything is joined with `&&` so if there is a test failure then the step is short cut to save on processing time.

0liver replied:
This is quite magical.

Piedone replied:
I see, interesting. Please document this because just by looking at it I would have a very hard time figuring out what's the point here. Also, try to structure it more for readability.

sarahelsaig replied:
You can't safely put comments inside the script section (because it gets collapsed during parsing) so I had to write a wall of text above it :( . I have also formatted the code as much as possible, but sed and readability are not something you usually say in the same sentence.

Piedone replied:
"sed"? :D

Yeah, I was meaning YAML comments anyway.

sarahelsaig replied:
[sed](https://en.wikipedia.org/wiki/Sed) = stream editor, a classic Unix program. It does text replacing, splicing, etc and can be piped. Basically vim for shell scripting. You know you are writing sh when your `cat | sed` 😿

Piedone replied:
Thanks, I hate it! :D

sarahelsaig committed:
Add further comment on SDK bug.

sarahelsaig committed:
What SDK version?

sarahelsaig committed:
Upgrade to Rnwood.Smtp4dev 3.1.4

sarahelsaig committed:
Comment out the SDK workaround, but explain.

sarahelsaig committed:
Retry on NPM troubles

sarahelsaig committed:
Fix CliProgram ".exe" double suffixing bug.

Piedone committed:
Code styling

Piedone commented on .github/workflows/analyze_and_test.yml:
```diff
@@ -0,0 +1,56 @@
+        dotnet --list-sdks
+        npm install pnpm -g
+        until dotnet build Lombiq.OSOCE.sln --no-incremental -warnaserror /p:TreatWarningsAsErrors=true /p:RunAnalyzersDuringBuild=true -nologo -consoleLoggerParameters:NoSummary -verbosity:quiet >analysis.log 2>&1; do
+          grep 'error MSB3073: The command "npm run dotnet-postclean --if-present"' analysis.log >/dev/null 2>/dev/null && echo npm retry && continue
```
We run a Gulp Extensions build first in TC. Might be needed.

sarahelsaig replied:
ok I will try

sarahelsaig committed:
Update .github/workflows/analyze_and_test.yml

Co-authored-by: Zoltán Lehóczky <zoltan.lehoczky@lombiq.com>

sarahelsaig committed:
Update .github/workflows/analyze_and_test.yml

Co-authored-by: Zoltán Lehóczky <zoltan.lehoczky@lombiq.com>

sarahelsaig committed:
Merge branch 'issue/OSOE-49' of https://github.com/Lombiq/Open-Source-Orchard-Core-Extensions into issue/OSOE-49

sarahelsaig committed:
Merge build and analyze steps, rename yml file.

sarahelsaig committed:
Add failure dumps and clean up workflow a bit.

sarahelsaig committed:
Try pre-building Lombiq.Gulp.Extensions

sarahelsaig committed:
Merge remote-tracking branch 'origin/dev' into issue/OSOE-49

sarahelsaig committed:
On any PR

sarahelsaig committed:
Fix docker setup after previous commit

sarahelsaig committed:
Fix actions/upload-artifact@v2 path

sarahelsaig committed:
Intentional test failure.

sarahelsaig committed:
Strip invalid path characters to please actions/upload-artifact@v2.

sarahelsaig committed:
Improve MakeFileSystemFriendly.

sarahelsaig committed:
Add missing badded character.

sarahelsaig committed:
Remove intentional error as actions/upload-artifact@v2 now works.

sarahelsaig committed:
unusing

sarahelsaig committed:
Rename shared config path environment variable.

sarahelsaig committed:
Move CommandExtensions.

sarahelsaig committed:
Add comments to yml.

sarahelsaig committed:
How about using "Password1!" for this too?

sarahelsaig committed:
Hopefully fix non-Docker SQL.

sarahelsaig committed:
Intentional test error in SQL test.

sarahelsaig committed:
Revert

sarahelsaig committed:
Rename CommandAsync to ExecuteAsync

sarahelsaig committed:
Update submodules

sarahelsaig committed:
Display setup validation errors.

sarahelsaig committed:
Unusing

sarahelsaig committed:
Simplify TryFindVersionAsync.

sarahelsaig committed:
Code cleanup

sarahelsaig committed:
Add comment about WebDriverManager.Net PR.

Piedone commented on .github/workflows/build_and_test.yml:
```diff
@@ -0,0 +1,61 @@
+      if: failure()
+      with:
+          name: ui-test-failure
+          path: test/**/FailureDumps
```
This also runs on unit test failure and causes a warning: https://github.com/Lombiq/Open-Source-Orchard-Core-Extensions/actions/runs/2333914401 Maybe better run unit and UI tests in separate tests?

sarahelsaig replied:
That won't help. The `if: failure()` will evaluate true if any preceding tests failed. See this example:
https://github.com/Lombiq/Open-Source-Orchard-Core-Extensions/runs/6459542404?check_suite_focus=true

Piedone replied:
Do you mean if any preceding step failed? This seems the be the case, [see docs](https://docs.github.com/en/actions/learn-github-actions/expressions#failure), but you can make steps depend on each other to make the UI testing and failure dump-creating step connected.

sarahelsaig replied:
How do you tell UI and unit tests apart anyway?

sarahelsaig replied:
I have found [this gem](https://docs.github.com/en/actions/learn-github-actions/expressions#hashfiles) and it seem to work. Doing `if: failure() && hashFiles('test/**/FailureDumps')` means upload-artifact only runs if there is something to dump.

Piedone replied:
Great then! For the first question, I suppose based on name only.

sarahelsaig replied:
> based on name only.

[shudder]

Piedone committed:
Typos, code styling

Piedone commented on .github/workflows/build_and_test.yml:
```diff
@@ -0,0 +1,61 @@
+      run: |
+        npm install pnpm -g
+        dotnet build src/Utilities/Lombiq.Gulp.Extensions/Lombiq.Gulp.Extensions.csproj --configuration Release
+        dotnet build Lombiq.OSOCE.sln --configuration Release -warnaserror /p:TreatWarningsAsErrors=true /p:RunAnalyzersDuringBuild=true -nologo -consoleLoggerParameters:NoSummary -verbosity:quiet
```
We should also build the NuGet solution. That can perhaps be a second job?

sarahelsaig committed:
Intentional Failure

sarahelsaig committed:
Conditional artifacts?

sarahelsaig committed:
oops

sarahelsaig committed:
Remove failure.

sarahelsaig committed:
Fix theme problems.

sarahelsaig committed:
Wake up, Neo. The Matrix has you...

sarahelsaig committed:
syntax

sarahelsaig committed:
You can't concatenate strings wtf

sarahelsaig committed:
Fix typo

sarahelsaig committed:
The operation was canceled.

sarahelsaig committed:
dont cd dot

sarahelsaig committed:
experiment

sarahelsaig committed:
Temporarily disable NuGetTestbed

sarahelsaig committed:
Add SetupShellCliCommandFactory

sarahelsaig committed:
Temporarily hard add editorconfig.

sarahelsaig committed:
Why CS1591

sarahelsaig committed:
CLSCompliant

sarahelsaig committed:
suppress CA1014

sarahelsaig committed:
format command in yml and -warnAsMessage:CS1591

sarahelsaig committed:
format parameter

sarahelsaig committed:
fix error code

sarahelsaig committed:
Update Lombiq.Tests.UI PackageReference

sarahelsaig committed:
Update Lombiq.Analyzers.

sarahelsaig committed:
Merge remote-tracking branch 'origin/issue/OSOE-49' into issue/OSOE-49

sarahelsaig committed:
Update BaseTheme and DataTables packagereference versions.

sarahelsaig committed:
Regenerated editorconfig?

sarahelsaig committed:
Fix malformed version codes.

sarahelsaig committed:
Move base theme test into BehaviorModuleTests.

sarahelsaig committed:
Update UI Testing Toolkit dependency.

sarahelsaig committed:
Update package reference versions.

sarahelsaig committed:
Fix test.

sarahelsaig committed:
Update package references.

sarahelsaig committed:
Add Lombiq.TrainingDemo to the nuget recipe.

sarahelsaig committed:
Update package references.

sarahelsaig committed:
Fully update Lombiq.OSOCE.NuGet.Web features list from Lombiq.OSOCE.Web

Piedone commented on NuGetTest/.editorconfig:
```diff
@@ -0,0 +1,287 @@
+# WARNING: Only edit this file in the Lombiq .NET Analyzers repository's folder. A copy of this file anywhere else will
```
Why is this file needed? We shouldn't have to copy it. If Analyzers adds it then it should be gitignored.

Piedone commented on NuGetTest/test/Lombiq.OSOCE.NuGet.Tests.UI/Tests/BehaviorModuleTests.cs:
```diff
@@ -19,4 +21,17 @@ public BehaviorModuleTests(ITestOutputHelper testOutputHelper)
             browser);
+
+    [Theory, Chrome]
+    public Task BaseThemeShouldWork(Browser browser) =>
```
Why did you move this from `BehaviorBaseThemeTests`? It was good there, it's not a module.

sarahelsaig replied:
`BehaviorBaseThemeTests` is too specific. If it's not good here, then I'll move it to `BehaviorThemeTests` (without the "Base").

Piedone replied:
Yeah, just a rename like that is good.

Piedone committed:
Code clean-up

Piedone commented:
You didn't address my feedback here.

sarahelsaig committed:
remove editorconfig

sarahelsaig committed:
Move theme test to BehaviorThemeTests.cs

sarahelsaig committed:
Update html-validate config

based on https://github.com/Lombiq/UI-Testing-Toolbox/pull/119#issuecomment-1132134078

sarahelsaig committed:
Fix syntax.

Piedone reviewed (approved):
Looks good but https://github.com/Lombiq/UI-Testing-Toolbox/pull/119#discussion_r877536815 affects this too.

sarahelsaig committed:
update submodules

sarahelsaig committed:
Revert everything that doesn't need to be async

in AtataFactory and WebDriverFactory.

sarahelsaig committed:
update DT submodule

sarahelsaig committed:
More neural language parser friendly comment in gitattributes.

sarahelsaig committed:
Backtick

sarahelsaig committed:
update submodules

sarahelsaig committed:
Merge remote-tracking branch 'origin/dev' into issue/OSOE-49

sarahelsaig committed:
Code cleanup.

sarahelsaig committed:
Update html-validate version.

sarahelsaig committed:
Merge remote-tracking branch 'origin/dev' into issue/OSOE-49

0liver commented on .github/workflows/build_and_test.yml:
```diff
@@ -0,0 +1,75 @@
+        docker exec -u 0 sql2019 bash -c 'mkdir /data; chmod 777 /data -R; chown mssql:root /data'
+    - name: Set up Azurite via Docker
+      run: |
+        docker pull mcr.microsoft.com/azure-storage/azurite &&
```
Why do you chain with `&&` here but not in the SQL server setup [above](https://github.com/Lombiq/Open-Source-Orchard-Core-Extensions/pull/41/files#diff-48c0ee97c53013d18d6bbae44648f7fab9af2e0bf5b0dc1ca761e18ec5c478f2R39)?

sarahelsaig replied:
By accident.

0liver commented on .github/workflows/build_and_test.yml:
```diff
@@ -0,0 +1,75 @@
+      # This step gets the list of all projects, selects the test projects (excluding the two test libraries) and then
+      # uses sed to create an inline bash script. What this script does:
+      # - Uses 'dotnet test -c Release --no-restore --no-build' to start up each project.
+      # - Before the test outputs '::group::project name' and after outputs '::endgroup::'. This tells GitHub that the
+      #   outputs should be placed into a collapsible region on the Actions web UI.
```
I needed some clarification on this comment, so here's what I came up with:
```suggestion
      # - The test output is wrapped by '::group::<project name>' before and '::endgroup::' after it. This tells GitHub
      #   to place each test's output into a collapsible region on the Actions web UI.
```

sarahelsaig replied:
Hmm, some nuance is lost in translation, but I guess it carries the gist better so I'm in favor of it.

0liver commented on .github/workflows/build_and_test.yml:
```diff
@@ -0,0 +1,75 @@
+      #   outputs should be placed into a collapsible region on the Actions web UI.
+      # - Each call is connected with '&&', so it stops on the first test failure to save computational resources.
+      #
+      # The script is evaluated by bash and the output is passed through grep to filter out some IPv6 related garbage
+      # data produced by chromedriver. The trouble is, this way the final error code will be grep's and not the test
+      # run's, so to get around this 'SUCCESS' or 'FAILURE' is written into the .ui_test_result file before the output
+      # is sent to grep. The final line sets the result according to what was written into the .ui_test_result file.
+      run: |
+        export LOMBIQ_UI_TESTING_TOOLBOX_SHARED_TEST_CONFIGURATION="$PWD/.github/TestConfig.json"
+        cd "${{matrix.directory}}"
+        (dotnet sln list |
+            grep '.Tests.' |
+            grep -v 'Lombiq.Tests.UI.csproj' |
+            grep -v 'Lombiq.Tests.csproj' |
+            sed -e 's/\\/\//g' \
+                -e '$atrue' \
+                -e 's/\(.*\)/echo "::group::\1" \&\&\ndotnet test -c Release --no-restore --no-build '"'\\1'"' \&\&\necho "::endgroup::" \&\&/' |
+            bash && (echo SUCCESS > .ui_test_result) || (echo FAIL > .ui_test_result)) |
+            grep -v '^Connection refused \[::ffff:127.0.0.1\]';
+        grep SUCCESS .ui_test_result
```
I'm seriously impressed.

0liver commented on NuGetTest/test/Lombiq.OSOCE.NuGet.Tests.UI/.htmlvalidate.json:
```diff
@@ -14,5 +14,16 @@
+    {
+      "form": {
+        "attributes": {
+          "method": [ "get", "post", "GET", "POST" ]
```
Is this now supposed to be unnecessary with the [latest version](https://gitlab.com/html-validate/html-validate/-/releases/v7.1.1)?

sarahelsaig committed:
Chaining.

sarahelsaig committed:
Update .github/workflows/build_and_test.yml

Co-authored-by: 0liver <0liver@users.noreply.github.com>

sarahelsaig committed:
Remove no longer needed configuration.

sarahelsaig committed:
Remove ImplicitUsings for the time being.

Commit:
Update Lombiq.UITestingToolbox

sarahelsaig committed:
test with latest issue/OSOE-49-suggestion

sarahelsaig committed:
update submodule

sarahelsaig committed:
Code cleanup.

sarahelsaig committed:
fix error

Piedone commented on Lombiq.OSOCE.sln:
```diff
@@ -122,6 +124,8 @@ Project("{2150E333-8FDC-42A3-9474-1A3956D46DE8}") = "Themes", "Themes", "{93A678
 EndProject
 Project("{FAE04EC0-301F-11D3-BF4B-00C04F79EFBC}") = "Lombiq.BaseTheme.Tests.UI", "src\Themes\Lombiq.BaseTheme\Lombiq.BaseTheme.Tests.UI\Lombiq.BaseTheme.Tests.UI.csproj", "{2BF13E3A-B54D-4352-812E-9A867A801FFE}"
 EndProject
+Project("{FAE04EC0-301F-11D3-BF4B-00C04F79EFBC}") = "Lombiq.HelpfulLibraries.Cli", "src\Libraries\Lombiq.HelpfulLibraries\Lombiq.HelpfulLibraries.Cli\Lombiq.HelpfulLibraries.Cli.csproj", "{F4B0F77A-4EF0-40A9-B1C0-9294B397BA36}"
```
Add a corresponding line to the Readme under "Included Projects".

sarahelsaig committed:
Amend Included Projects in the root Readme.

sarahelsaig committed:
Unify PNPM docs.

sarahelsaig committed:
update submodule

sarahelsaig committed:
OrchardCoreConfiguration improvements.

sarahelsaig committed:
Merge remote-tracking branch 'origin/dev' into issue/OSOE-49

# Conflicts:
#	src/Modules/Lombiq.Hosting.Azure.ApplicationInsights

sarahelsaig committed:
dummy commit

sarahelsaig committed:
Revert "dummy commit"

This reverts commit 777ad646beef35f57829d6568a8a6055b06c08b3.

Commit:
Update submodules to dev after merge
